hypothesis-python/src/hypothesis/core.py

```diff
@@ -1012,6 +1007,23 @@ def run(data):
                 if parts := getattr(data, "_stateful_repr_parts", None):
                     self._string_repr = "\n".join(parts)
 
+                # if is_final, we will append to "representation" in the final
+                # failing example in run_engine. Don't duplicate by also appending
+                # here. We prefer to append in run_engine because that respects
+                # the order in which other notes were added around draws.
+                if TESTCASE_CALLBACKS and not is_final:
+                    printer = RepresentationPrinter(context=context)
+                    for name, value in data._observability_args.items():
+                        if name.startswith("generate:Draw "):
+                            try:
+                                value = data.provider.realize(value)
+                            except BackendCannotProceed:
+                                value = "<backend failed to realize symbolic>"
+                            printer.text(f"\n{name.removeprefix('generate:')}: ")
+                            printer.pretty(value)
```

Comment:
I think here we'd prefer to have `printer.pretty()` handle the realization (or not), passing in `data.provider.avoid_realization`.  I'd also do it here unconditionally, and drop the notes from failing examples.

That takes out the coverage-based explanations, which is fine; we're on track to do better in HypoFuzz anyway.

Reply:
We're after the test case has finished, so I don't think there's any need to respect `data.provider.avoid_realization`. Though possibly we can just append the symbolic to the string since we realize in `self._string_repr = data.provider.realize(self._string_repr)` lower down, though that fails if *any* of its components fail, so checking each individually is more likely to succeed.

>  I'd also do it here unconditionally, and drop the notes from failing examples.

I think this will end up with `note("a")` not appearing in observability's `representation` arg for failing examples, even though hypothesis prints it. Seems pretty arbitrary; I'd expect `representation` to basically be "what hypothesis prints"

Reply:
Hmm, I don't share that expectation; to me the natural `representation` is the arguments and all the draws, and then notes are outputs from rather than inputs to the test.

Reply:
but you'd still want `representation` to have arguments annotated with the explain phase? That seems like a test output to me

Reply:
It's a little blurry, yeah, but that is what I want.